### PR TITLE
Write & Read State Machine To RocksDB

### PR DESCRIPTION
**Pull request by redixhumayun**
## What This PR Does

* Routes all state machine reads and writes for forward indexes to RocksDB
* State machine reads and writes to reverse indexes remain in memory
* Adds significant test coverage
* `RocksDB` uses `OptimisticTransactionDB` to provide transaction functionality

### CODE REVIEW

File: src/task_allocator/planner/load_aware_distributor.rs [hunk 1]
```diff
@@ -146,26 +147,34 @@ impl LoadAwareDistributor {
...
-                        .or_default()
-                        // use `Reverse` here to make it a min-heap
-                        .push(Reverse(ExecutorLoad {
-                            executor_id: executor_id.clone(),
-                            running_task_count,
-                        }));
+            let executor = self
+                .shared_state
+                .state_machine
+                .get_from_cf(StateMachineColumns::executors, executor_id)
```
Comment (diptanu) [inline src/task_allocator/planner/load_aware_distributor.rs:154]: Can this be a method in the state machine which returns the deserailzied executor?

File: src/task_allocator/planner/load_aware_distributor.rs [hunk 2]
```diff
@@ -353,42 +362,53 @@ mod tests {
...
         let shared_state = App::new(config, None).await.unwrap();
         shared_state.initialize_raft().await.unwrap();
         let _coordinator = crate::coordinator::Coordinator::new(shared_state.clone());
-        let sm = shared_state.indexify_state.read().await;
 
         // get tasks from the state
-        let tasks: HashSet<TaskId> = sm.tasks.values().map(|t| t.id.clone()).collect();
+        let serialized_tasks = shared_state
+            .state_machine
+            .get_all_rows_from_cf(StateMachineColumns::tasks)
```
Comment (diptanu) [inline src/task_allocator/planner/load_aware_distributor.rs:376]: Same here

**Review by diptanu — CHANGES_REQUESTED** [review]

File: src/state/store/mod.rs [hunk 3]
```diff
@@ -77,67 +84,42 @@ pub struct StateMachineStore {
...
-
-        Ok(Snapshot {
-            meta,
-            snapshot: Box::new(Cursor::new(indexify_state_json)),
-        })
-    }
+#[derive(strum::Display, strum::EnumIter)]
+#[allow(non_camel_case_types)]
+pub enum StateMachineColumns {
+    executors,
```
[the remark this answers is not in the document]
Comment (redixhumayun) [reply src/state/store/mod.rs]: Do we want these all to be PascalCase?

File: src/task_allocator/planner/load_aware_distributor.rs (hunk 1 above)
Comment (redixhumayun) [reply src/task_allocator/planner/load_aware_distributor.rs:154]: I've added a generic deserialization chained method here

File: src/task_allocator/planner/load_aware_distributor.rs (hunk 2 above)
Comment (redixhumayun) [reply src/task_allocator/planner/load_aware_distributor.rs:376]: I've added a generic deserialization chained method here

File: src/state/store/mod.rs [hunk 4]
```diff
@@ -291,7 +442,12 @@ impl RaftStateMachine<TypeConfig> for StateMachineStore {
                 EntryPayload::Blank => {}
                 EntryPayload::Normal(req) => {
                     change_events.extend(req.new_state_changes.clone());
-                    sm.apply(req.clone());
+
+                    if let Err(e) = sm.apply_state_machine_updates(req.clone(), &self.db) {
```
Comment (diptanu) [inline src/state/store/mod.rs:446]: No need to clone

**Review by diptanu — CHANGES_REQUESTED** [review]

**Pull request merged**